data/mapper/object.go

```diff
@@ -83,182 +82,221 @@ func (am *ObjectMapperFactory) NewExpr(value string) (expression.Expr, error) {
 	}
 }
 
-func hasForeach(val map[string]interface{}) bool {
-	for foreachK, _ := range val {
-		if strings.HasPrefix(foreachK, FOREACH) {
-			return true
+func (f *foreachExpr) addFields(fields map[string]interface{}, exprF expression.Factory) (err error) {
+	for key, value := range fields {
+		if key == "=" {
+			if value == "$loop" {
+				f.assign = &assignAllExpr{}
+			} else {
+				f.assign, err = newExpr(value, exprF)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			if f.fields == nil {
+				f.fields = make(map[string]expression.Expr)
+			}
+			f.fields[key], err = NewObjectMapper(value, exprF)
+			if err != nil {
+				return err
+			}
 		}
 	}
-	return false
+	return err
 }
 
-func handlerArrayElement(element interface{}, exprF expression.Factory, inputScope data.Scope) (interface{}, error) {
-	//Only handle object mapping
-	switch element.(type) {
-	case map[string]interface{}:
-		return handleObjectMapping(element, exprF, inputScope)
-	default:
-		return getExpressionValue(element, exprF, inputScope)
+func newExpr(path interface{}, exprF expression.Factory) (expression.Expr, error) {
+	if isExpr(path) {
+		return exprF.NewExpr(path.(string)[1:])
+	} else {
+		return expression.NewLiteralExpr(path), nil
+	}
+}
+
+func newForeachExpr(foreachpath string, exprF expression.Factory) (*foreachExpr, error) {
+	foreach := &foreachExpr{}
+	foreachpath = strings.TrimSpace(foreachpath)
+	if strings.HasPrefix(foreachpath, FOREACH) && strings.Contains(foreachpath, "(") && strings.Contains(foreachpath, ")") {
+		paramsStr := foreachpath[strings.Index(foreachpath, "(")+1 : strings.LastIndex(foreachpath, ")")]
+		sourceIdx := strings.Index(paramsStr, ",")
+		if sourceIdx <= 0 {
+			finalExpr, err := exprF.NewExpr(strings.TrimSpace(paramsStr))
+			if err != nil {
+				return nil, err
+			}
+			foreach.sourceFrom = finalExpr
+		} else {
+
+			finalExpr, err := exprF.NewExpr(strings.TrimSpace(paramsStr[:sourceIdx]))
+			if err != nil {
+				return nil, err
+			}
+			foreach.sourceFrom = finalExpr
+
+			if len(paramsStr) > sourceIdx+1 {
+
+				afterLoopNameParamStr := strings.TrimSpace(paramsStr[sourceIdx+1:])
+				loopNameIdx := strings.Index(afterLoopNameParamStr, ",")
+				if loopNameIdx >= 0 {
+					foreach.scopeName = strings.TrimSpace(afterLoopNameParamStr[:loopNameIdx])
+				} else {
+					foreach.scopeName = afterLoopNameParamStr
+					return foreach, nil
+				}
+
+				if len(afterLoopNameParamStr) > loopNameIdx+1 {
+					filter := strings.TrimSpace(afterLoopNameParamStr[loopNameIdx+1:])
+					if len(filter) > 0 {
+						//create new filter expression
+						filterExpr, err := exprF.NewExpr(filter)
+						if err != nil {
+							return nil, fmt.Errorf("create foreach filtering expression error: %s", err.Error())
+						}
+						foreach.filterExpr = filterExpr
+					}
+				}
+			}
+		}
+	}
+	return foreach, nil
+}
+
+func (obj *ObjectMapper) Eval(scope data.Scope) (value interface{}, err error) {
+	defer func() {
```

Comment:
Just a heads up that defer is expensive, we should make sure that we don't have too many of these nested

Reply:
@fm-tibco  We have defer func at EvalActivity at TaskInst but we don't have defer func for trigger handler.   Do you think we should keep here or remove it from here and it at trigger handler?

Reply:
Lets move it to the trigger handler... this will help catch other issues as well.

Reply:
Sure.  I removed defer here and will create another PR at flow repo to catch defer in the handler.

Reply:
My fault.  I will add to this PR. Its part of core :)